File: ExercismFSharp/ExercismFSharp/RunLengthEncoding.fs
module RunLengthEncoding

open System.Text.RegularExpressions

let encode input =
    Regex.Matches(input, "(.)\1*")
    |> Seq.cast
    |> Seq.map (fun (m : Match) ->
        let result = m.Groups.[0].ToString()
        let length = result |> Seq.length
        let char = result.[0]

        match length with
        | 1 -> sprintf "%c" char
        | _ -> sprintf "%i%c" length char
    )
    |> Seq.reduce (sprintf "%s%s")

let decode input =
    Regex.Matches(input, "(\d*)(.)")
    |> Seq.cast
    |> Seq.map (fun (m : Match) ->
        let numOcc = m.Groups.[1].ToString()
        let letter = m.Groups.[2].ToString()

        match numOcc with
        | "" -> letter
        | _  ->
            Seq.init (numOcc |> int) (fun _ -> letter)
            |> Seq.reduce (sprintf "%s%s")
    )
    |> Seq.reduce (sprintf "%s%s")

File: Exercism/python/run-length-encoding/run_length_encoding.py
import re

from functools import reduce


def decode(string):
    groups = [decode_single_match(m) for m in re.compile(r"(\d+)?(.)").finditer(string)]
    return reduce(lambda x, y: x + y, groups, "")


def decode_single_match(match):
    if not match.group(1):
        return match.group(2)

    return match.group(2) * int(match.group(1))


def encode(string):
    groups = [m.group() for m in re.compile(r"(.)\1*").finditer(string)]
    return reduce(lambda x, y: x + encode_single_value(y), groups, "")


def encode_single_value(string):
    if len(string) == 1:
        return string

    return str(len(string)) + string[0]
